### ISSUE TITLE
Scheduler doesn't match async-await timing

### ISSUE THREAD

**Issue by machty** [machty/ember-concurrency#54]
At present, all `perform()`s are deferred to the `actions` queue, which occasionally produces different results from putting the same code in an `Ember.on()` function, e.g.

``` js
export default Ember.Component.extend({
  afterInit: Ember.on('init', function() {
    // runs synchronously during/after `init`
    this.set('wat', 'lol');
  }),
  // vs
  afterInitTask: task(function * () {
    // happens later, on the `actions` queue
    this.set('wat', 'lol');
  }).on('init'),
});
```

I noticed this was causing undesired animations to occur with Liquid Fire which were fixed by changing the code back to user `Ember.on('init')`. I don't know what the solution is now, but given that most libraries assume that logic in hooks is performed synchronously when the hook is called, ember-concurrency should probably make it possible to perform logic synchronously. Or perhaps e-c can maintain a microqueue that it trampoline-flushes every time `.perform()` is called.

**Comment by machty** [comment]
Here's the plan: ember-concurrency's gonna get a new scheduler to bring it closer in line with how async-await timing works.

```js
    async function wat() {
      console.log(1);
      await null;
      console.log(3);
    }

    wat();
    console.log(2);
```

This produces 1 2 3

```js
    myTask: task(function * () {
      console.log(1);
      yield null;
      console.log(3);
    }),

    myTask.perform();
    console.log(2);
```

Today, this produces 2 1 3 because `.perform()` doesn't immediately (synchronously) start executing your task until the actions run loop queue. This is 1) inconsistent w async-await semantics, which is bad because ember-concurrency strives to match async-await semantics, but with the addition of cancelation, and 2) problematic with `.on('init')` or any time you perform a task from some third party API hook that expects some side effect to be performed before control is returned to the API.

So I'm going to change the scheduler to make the second example above print 1 2 3.

**Comment by machty** [comment]
## To Sync, or not To Sync

After talking over w the core team, I don't think we can/should make all tasks begin execution synchronously by default. Instead, we should make it possible to opt into synchronous execution.

The reason we can't make synchronous start the default is because we can't guarantee synchronous execution in the case of restartable tasks.

```js
   foo: task(function * () {
     yield timeout(100);
     this.get('foo').perform(); // what happens here?
   }).restartable()
```

What happens in the above case assuming we have synchronous start? `foo` gets performed _while a TaskInstance of `foo` is actively executing_ (its internal generator function iterator is literally mid-execution), but we'd have to start a new TaskInstance of `foo` and execute it immediately. If we let these TaskInstances of `foo` overlap, then we've broken the guarantee that `restartable` reliably constrains concurrency, albeit for a small slice of execution time, but it's enough of a problem that I don't even want to open the door to the kinds of subtle, hair-pulling-out errors that could occur if e-c even slightly broke its promises about task modifiers and concurrency constraints.

Note that above code snippet could be modified to add some indirection and the problem would still be the same; if `foo` calls other methods/tasks and ultimately something happens that ends up calling `foo.perform()` again within the same call stack, the same problems occur.

So, for these reasons, we can't make sync start the default; defaulting to always-async sidesteps an entire class of messy Zalgo-y problems.

## Alternative A: expose `.performSync()` method

We make sync start opt-in by adding a `performSync()` method on Task. This would allow users/libraries within timing/sync-dependent hooks to have the guarantees that the task will start running synchronously, much like async functions do.

If we go this route, we probably would NOT need to add a `perform-sync` Handlebars helper, since I can't think of a use case where it would matter to Handlebars-initiated performs whether a task began synchronously; rather, it's only timing-dependent JS code where sync start would ever be desirable.

Calling `.performSync()` on restartable tasks would throw an error. (Should it also throw for `.drop` / `enqueue`? Maybe we should start off disallowing it with ALL task modifiers and consider relaxing constraints in a future version.)

## Alternative B: add `.syncStart()` Task Modifier

Add a new task modifier called `.syncStart()` that changes the behavior so that calling `perform()` starts synchronously executing the task. Combining `.syncStart()` with other task modifiers would throw an error.

I probably prefer this approach more because:

1. It fails as early as possible rather than waiting for someone to call `.performSync()`
2. It's jells nicely with the upcoming [toFunction()](https://github.com/machty/ember-concurrency/issues/80#issuecomment-253306367) feature; toFunction() will by default produce functions that call `perform()` with async start by default, but could be made to behave with async function semantics by doing `task(function * () {}).syncStart().toFunction()`.

**Comment by machty** [comment]
After some _more_ conversations with ember core I'm giving synchronous performs a go; progress can be followed at #100.

**Comment by machty** [comment]
Merged into master, coming out in 0.8.0. Didn't end up doing the syncStart route.
